### PR TITLE
Load state dict post hook

### PR DESCRIPTION
**Pull request by rohan-varma**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #75424
* __->__ #75422

Adds load_state_dict_post_hook. Proposal: https://github.com/pytorch/pytorch/issues/75287

Went with no args except for module optionally for now. Additional args can be
added on use case basis.

A few notes:
- Similar to state_dict post hook, we run it after this module and all submodule children have been loaded.
- Added try/finally to ensure the hook runs even if load_state_dict throws.

Differential Revision: [D35439200](https://our.internmc.facebook.com/intern/diff/D35439200/)

### CODE REVIEW

File: torch/nn/modules/module.py [hunk 1]
```diff
@@ -1537,12 +1562,18 @@ def load_state_dict(self, state_dict: 'OrderedDict[str, Tensor]',
...
-                    load(child, prefix + name + '.')
+            try:
+                local_metadata = {} if metadata is None else metadata.get(prefix[:-1], {})
+                module._load_from_state_dict(
+                    state_dict, prefix, local_metadata, True, missing_keys, unexpected_keys, error_msgs)
+                for name, child in module._modules.items():
+                    if child is not None:
+                        load(child, prefix + name + '.')
+            finally:
+                # Note that post hooks for a module are called after self _and_ all
```
Comment (awgu) [inline torch/nn/modules/module.py:1573]: nit: duplicate `all`

File: torch/nn/modules/module.py [hunk 2]
```diff
@@ -1407,6 +1410,27 @@ def _register_load_state_dict_pre_hook(self, hook, with_module=False):
         self._load_state_dict_pre_hooks[handle.id] = hook
         return handle
 
+    def _register_load_state_dict_post_hook(self, hook, with_module=False):
```
Comment (zhaojuanmao) [inline torch/nn/modules/module.py:1413]: curious why "with_module" option is needed?

**Review by malfet — APPROVED** [review]

**Review by malfet — CHANGES_REQUESTED** [review]
Not requesting any changes, just rewoking my review as PR has not been approved internally (nor publically)

File: torch/nn/modules/module.py (hunk 2 above)
Comment (jbschlosser) [reply torch/nn/modules/module.py:1432]: Looks like this was done to match `_register_load_state_dict_pre_hook()`, where the `with_module` flag was added later on.

Comment (albanD) [reply torch/nn/modules/module.py:1432]: Note that we wrote a full issue on how we want these hooks to happen with Rohan. This PR is outdated wrt that and should be updated to match.

**Comment by rohan-varma** [comment]
Closing in favor of new API to be implemented as part of addressing https://github.com/pytorch/pytorch/issues/75287

**Pull request closed without merging**